shared/src/main/scala/mlscript/TyperDatatypes.scala

```diff
@@ -717,41 +717,41 @@ abstract class TyperDatatypes extends TyperHelpers { Typer: Typer =>
     }
     def lcg(pol: Opt[Bool], first: ST, rest: Ls[Opt[ST]])
       (implicit prov: TypeProvenance, ctx: Ctx)
-        : (ST, Ls[(Opt[Bool], ST)], Ls[Ls[Opt[ST]]]) = first match {
+        : (ST, Ls[(Opt[Bool], ST)], Ls[Ls[Opt[ST]]]) = first.unwrapProxies match {
       case a: FunctionType =>
-        val (lhss, rhss) = rest.map {
+        val (lhss, rhss) = rest.map(_.map(_.unwrapProxies) match {
```

Comment:
That's a bit crazy. At every iteration you map the entire `rest` and apply `unwrapProxies` everywhere. It's very inefficient (quadratic) and also redundant/unnecessary.

Reply:
Ok I read too fast, it's ok. But you can make the improvement discussed in the meeting.